### PR TITLE
fix syntax error in is_bfloat16_supported

### PR DESCRIPTION
**Pull request by puririshi98**
fixes prior syntax error from PR @ngimel

### CODE REVIEW

**Comment by ngimel** [comment]
There doesn't seem to be any tests testing `torch.autocast` with `dtype` argument, and in particular, `dtype` bfloat16, which would have caught this error. Can you please add them?

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
yeah would be good to know why no tests caught this

**Comment by puririshi98** [comment]
So I added the tests in but I find an error now:

**Comment by puririshi98** [comment]
So I added in the tests and now get the following issue:

**Comment by puririshi98** [comment]
So I added in the tests for bfloat16
But I get the following error when running it locally on a A100:
```
ERROR: test_autocast_torch_bf16 (__main__.TestCuda)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 1075, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 1075, in wrapper
    method(*args, **kwargs)
  File "test_cuda.py", line 2797, in test_autocast_torch_bf16
    self._run_autocast_outofplace(op, args, torch.bfloat16)
  File "test_cuda.py", line 2721, in _run_autocast_outofplace
    output = getattr(module, op)(*args, **add_kwargs)
RuntimeError: getCudnnDataTypeFromScalarType() not supported for BFloat16
```

Which is caused by the following cpp code in the ATEN backend.
```
namespace at { namespace native {

cudnnDataType_t getCudnnDataTypeFromScalarType(const at::ScalarType dtype) {
  if (dtype == at::kFloat) {
    return CUDNN_DATA_FLOAT;
  } else if (dtype == at::kDouble) {
    return CUDNN_DATA_DOUBLE;
  } else if (dtype == at::kHalf) {
    return CUDNN_DATA_HALF;
  }
  std::string msg("getCudnnDataTypeFromScalarType() not supported for ");
  msg += toString(dtype);
  throw std::runtime_error(msg);
}
```
I am wondering why BFLOAT16 was not previously implemented for CUDNN and if that is something I should add in or something I should work around? When I grep for CUDNN_DATA_* I find their definitions in https://github.com/puririshi98/pytorch/blob/clean_err_msg_bf/caffe2/core/common_cudnn.h#L119-L188

**Comment by ngimel** [comment]
Convolutions don't support bfloat16, you'd need to add a test that runs matmuls only. You'd also need to catch an error in the test if bfloat16 is not supported.

**Comment by puririshi98** [comment]
> Convolutions don't support bfloat16, you'd need to add a test that runs matmuls only. You'd also need to catch an error in the test if bfloat16 is not supported.

Okay thanks for the explanation. Will do

**Comment by ngimel** [comment]
If you put a skip like this
```
@unittest.skipIf(not TEST_BF16, 'BFloat16 operations not available')
    def test_autocast_nn_bf16(self):
```
it would mean that this test will never run in our OSS CI, because our OSS CI doesn't have amperes, and thus wouldn't be able to catch even syntax errors that you had before. Instead, you shouldn't skip the full test, but, depending on `is_bf16_supported`, either compare the results with `self.assertEqual` (or whatever is used now), or make sure that the expected error is raised with `self.assertRaisesRegex`.

**Comment by puririshi98** [comment]
> If you put a skip like this
> 
> ```
> @unittest.skipIf(not TEST_BF16, 'BFloat16 operations not available')
>     def test_autocast_nn_bf16(self):
> ```
> 
> it would mean that this test will never run in our OSS CI, because our OSS CI doesn't have amperes, and thus wouldn't be able to catch even syntax errors that you had before. Instead, you shouldn't skip the full test, but, depending on `is_bf16_supported`, either compare the results with `self.assertEqual` (or whatever is used now), or make sure that the expected error is raised with `self.assertRaisesRegex`.

well I left the previous fp16 test untouched and the new bf16 will only run in the case that bf16 is supported. I assume this is equivalent?

**Comment by ngimel** [comment]
No it's not equivalent because in this case you skip whole test. The goal is to try to run the test and either make sure autocast works (if hardware supports bfloat16) or trying to run bfloat16 autocast on an unsupported hardware throws expected error (it's not the same as `is_bfloat16_available` returning False)

**Comment by ngimel** [comment]
Ok, this looks good, let's see what CI says.

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -82,7 +82,8 @@ def is_bf16_supported():
     r"""Returns a bool indicating if the current CUDA device supports dtype bfloat16"""
     cu_vers = torch.version.cuda
     if cu_vers is not None:
-        cuda_maj_decide = int(cu_vers.split(',')[0]) >= 11
+        cuda_maj_decide = int(cu_vers.split('.')[0]) >= 11
```
Comment (jstjohn) [inline torch/cuda/__init__.py:85]: I opened an issue for this here https://github.com/pytorch/pytorch/issues/64232. Is the ',' ever needed in some other locale, or was this just a bug?

Comment (jstjohn) [reply torch/cuda/__init__.py:85]: This exact change got it working for me, but I'm in the US so I wasn't sure if this was a battle of the decimal character or something :)

Comment (ngimel) [reply torch/cuda/__init__.py:85]: No, just plain bug.

File: test/test_cuda.py [hunk 2]
```diff
@@ -2784,6 +2787,27 @@ def test_autocast_torch_fp16(self):
...
+            for op_with_args in self.autocast_lists.torch_fp16:
+                skip_test = False
+                op, args = op_with_args[0], op_with_args[1]
+                if len(op_with_args) == 3:
+                    skip_test = op_with_args[2]  # TEST_WITH_ROCM
+                should_error_from_not_implemented = 'cudnn' in op or 'prelu' in op or 'thnn' in op \
+                    or 'fused' in op or 'gru' in op or op == '_thnn_fused_lstm_cell' or op == 'lstm_cell'
+                if not skip_test:
+                    if should_error_from_not_implemented:
+                        with self.assertRaises(RuntimeError,msg=str(op) + ' should not be supported for bfloat16!'):
```
Comment (ngimel) [inline test/test_cuda.py:2802]:
```suggestion
                        with self.assertRaises(RuntimeError, msg=str(op) + ' should not be supported for bfloat16!'):
```

**Pull request closed without merging**